routers/web/repo/tree.go

```diff
@@ -52,3 +53,30 @@ func isExcludedEntry(entry *git.TreeEntry) bool {
 
 	return false
 }
+
+func Tree(ctx *context.Context) {
+	treePath := ctx.PathParam("*")
+	var refFullName git.RefName
+	switch ctx.FormTrim("ref_type") {
+	case "branch":
+		refFullName = git.RefNameFromBranch(ctx.FormTrim("ref_name"))
+	case "tag":
+		refFullName = git.RefNameFromTag(ctx.FormTrim("ref_name"))
+	default:
+		refFullName = git.RefName(ctx.FormTrim("ref_name"))
+	}
```

Comment:
I think you need to figure out a fact first: we already used ref + path a lot in code, but there is no such code in code base, then there must be something with the newly written code.

Reply:
There is similar code as below which is committed by you 5 hours ago. https://github.com/go-gitea/gitea/commit/81352542fd29d1809bd483dec60849ec39d84388#diff-33ef19b39f07d3b142a34459f94663b18a5c5ae834c55d11efdb09745988c1c4R813-R826 but that's an internal function, maybe it can be public one so it can be reused.

```
func repoRefFullName(shortName string, typ RepoRefType) git.RefName {
	switch typ {
	case RepoRefBranch:
		return git.RefNameFromBranch(shortName)
	case RepoRefTag:
		return git.RefNameFromTag(shortName)
	case RepoRefCommit:
		return git.RefNameFromCommit(shortName)
	default:
		setting.PanicInDevOrTesting("Unknown RepoRefType: %v", typ)
		return git.RefNameFromBranch("main") // just a dummy result, it shouldn't happen
	}
}
```

Reply:
Why the ref type and ref name haven't been a part of the request path it's more clear with query parameters than path parameters like your previous review comment. `context.RepoRefByType` has many guesses logic. And since it's an Ajax request, the URL will not be exposed to the user directly.

Reply:
> Why the ref type and ref name haven't been a part of the request path it's more clear with query parameters than path parameters like your previous review comment. `context.RepoRefByType` has many guesses logic.

No, `RepoRefByType` doesn't guess unless you make it guess. `RepoRefByType` is the right one to call.

Reply:
And your code:

```
default:
		refFullName = git.RefName(ctx.FormTrim("ref_name"))
```

You made it support `refs/pull/1` here? I do not think it is right to do so.

Do not re-inventing the wheel

Reply:
> There is similar code as below which is committed by you 5 hours ago. [8135254#diff-33ef19b39f07d3b142a34459f94663b18a5c5ae834c55d11efdb09745988c1c4R813-R826](https://github.com/go-gitea/gitea/commit/81352542fd29d1809bd483dec60849ec39d84388#diff-33ef19b39f07d3b142a34459f94663b18a5c5ae834c55d11efdb09745988c1c4R813-R826) but that's an internal function, maybe it can be public one so it can be reused.

No, that's different, if you make it public, it is an abuse.

Reply:
I mean `RepoRefByType` will guess the branch/tag name.

Reply:
The rule is simple:

Use `RefWebLinkPath` to generate a link and use it in frontend, and backend uses `RepoRefByType` to parse that link for route handlers to get correct ref and treepath.

No need to (and shouldn't) manually write any new parser or converter.

Reply:
4890434
I created a new function `RefNameFromTypeAndShortName` in git module, maybe `func repoRefFullName(shortName string, typ RepoRefType) git.RefName` could be replaced by that one. Then the `RepoRefType` could be replaced by `git.RefType`

Reply:
> I mean `RepoRefByType` will guess/detect the first matching branch/tag name.

No, it doesn't guess, I think it is right unless you could show a real bad case.